cilium/CFP-34577-per-service-lb-algorithm-selection.md

```diff
@@ -0,0 +1,144 @@
+
+Similar requests that failed to gain traction:
+
+1. [cilium\#8723](https://github.com/cilium/cilium/issues/8723) \- algorithm selection in service annotation  
+2. [cilium\#26264](https://github.com/cilium/cilium/issues/26264) \- more load-balancing algorithms
+
+`kube-router` lets users [select a load-balancing algorithm based on service annotation](https://github.com/cloudnativelabs/kube-router/blob/v2.1.3/docs/user-guide.md#load-balancing-scheduling-algorithms).  
+At the time of this writing, there are [14 algorithms](https://keepalived-pqa.readthedocs.io/en/latest/scheduling_algorithms.html) to choose from.
+
+## Goals
+
+1. Support multiple load-balancing algorithms on a single node
+
+## Non-Goals
+
+1. Add support for new (other than `maglev`/`random`) load-balancing algorithms
+
+## Proposal
+
+Add a new value for `bpf-lb-algorithm` and a new `service.cilium.io/scheduler` annotation which will initially accept `maglev` or `random` values.
+
+### Overview
+
+1. Rename `lb{4,6}_select_backend_id` functions to include algorithm name  
+   [code](https://github.com/cilium/cilium/blob/v1.16.0/bpf/lib/lb.h#L1310-L1332)  
+2. Compile eBPF with `LB{4,6}_MAGLEV_MAP_OUTER`  
+   (default or behind a new value of `bpf-lb-algorithm`)  
+   [code](https://github.com/cilium/cilium/blob/v1.16.0/bpf/lib/lb.h#L78-L94)  
+3. Add new `flag` to `lb{4,6}_service`  
+   [code](https://github.com/cilium/cilium/blob/v1.16.0/bpf/lib/common.h#L1064-L1065)  
+4. Pass the flag based on `service.cilium.io/scheduler` annotation to `lb{4,6}_service`  
+5. Diverge load-balancing code paths based on the flag set in `lb{4,6}_service`
+
+#### Pros
+
+1. Slow rollouts \- a different algorithm can be selected/tested without affecting the entire cluster  
+2. Extensible:  
+   1. A new algorithms can be introduced without affecting current behavior (opt-in)  
+   2. It is possible to achieve a feature-parity with `envoy`/`kube-router` in the future
+
+#### Cons
+
+1. Adding/Removing annotation to a service may be hard to handle (without disruption)  
+2. Agent would need to support all LB algorithms at the same time  
+   When compared to uniform-only clusters. It will have:  
+   1. Slower eBPF compilation  
+   2. Slower startup due to map precomputation  
+   3. More BPF maps need to be allocated
+
+### Default behavior
+
+To avoid disruption during the cilium version upgrade, `bpf-lb-algorithm` will specify the default LB algorithm for services without `service.cilium.io/lb-algorithm` annotation.
+
+##### Alternative
+
+Add a new value, `bpf-lb-algorithm=all`, which will switch from the current behavior (of a single LB algorithm per cluster) to the one based on service annotations. In this case, the LB algorithm will default to `random` and could be changed to `maglev` only through setting `service.cilium.io/lb-algorithm=maglev` annotation.
```

Comment:
@brb @borkmann are you approving to implement the main proposal or the alternative?

If not the alternative, maybe we can clarify that, for example:
```suggestion
##### Alternative (not selected for implementation)

Add a new value, `bpf-lb-algorithm=all`, which will switch from the current behavior (of a single LB algorithm per cluster) to the one based on service annotations. In this case, the LB algorithm will default to `random` and could be changed to `maglev` only through setting `service.cilium.io/lb-algorithm=maglev` annotation.
```

Reply:
donne